File: packages/sdk/go/bugwatch.go
package bugwatch

import (
	"context"
	"fmt"
	"log"
	"sync"
)

var (
	globalClient *BugwatchClient
	globalMu     sync.RWMutex
)

// Init initializes the global Bugwatch client with the given options.
// Must be called before using any package-level capture functions.
func Init(options *BugwatchOptions) error {
	client, err := NewClient(options)
	if err != nil {
		return err
	}

	globalMu.Lock()
	globalClient = client
	globalMu.Unlock()

	return nil
}

// getClient returns the global client, or nil if not initialized.
func getClient() *BugwatchClient {
	globalMu.RLock()
	defer globalMu.RUnlock()
	return globalClient
}

// CaptureException captures an error using the global client.
// Returns the event ID, or an empty string if the client is not initialized.
func CaptureException(err error) string {
	client := getClient()
	if client == nil {
		return ""
	}
	return client.CaptureException(err)
}

// CaptureMessage captures a message at the given level using the global client.
// Returns the event ID, or an empty string if the client is not initialized.
func CaptureMessage(msg string, level Level) string {
	client := getClient()
	if client == nil {
		return ""
	}
	return client.CaptureMessage(msg, level)
}

// CaptureExceptionWithContext captures an error with context enrichment
// using the global client. Returns the event ID.
func CaptureExceptionWithContext(ctx context.Context, err error) string {
	client := getClient()
	if client == nil {
		return ""
	}
	return client.CaptureExceptionWithContext(ctx, err)
}

// RecoverAndCaptureException recovers from a panic and captures it as an
// exception using the global client. Intended to be used with defer:
//
//	defer bugwatch.RecoverAndCaptureException()
func RecoverAndCaptureException() {
	if r := recover(); r != nil {
		client := getClient()
		if client == nil {
			return
		}

		var err error
		switch v := r.(type) {
		case error:
			err = v
		case string:
			err = fmt.Errorf("%s", v)
		default:
			err = fmt.Errorf("panic: %v", v)
		}

		client.mu.RLock()
		debug := client.options.Debug
		client.mu.RUnlock()

		if debug {
			log.Printf("[Bugwatch] Recovered panic: %v", err)
		}

		client.CaptureException(err)
	}
}

// Flush flushes any pending events in the global client.
func Flush() {
	client := getClient()
	if client == nil {
		return
	}
	client.Flush()
}

// Close flushes and closes the global client.
func Close() {
	client := getClient()
	if client == nil {
		return
	}
	client.Close()

	globalMu.Lock()
	globalClient = nil
	globalMu.Unlock()
}

File: packages/sdk/java/core/src/main/java/dev/bugwatch/Bugwatch.java
package dev.bugwatch;

/**
 * Static facade for the Bugwatch SDK.
 * <p>
 * Initialize once at application startup:
 * <pre>{@code
 * Bugwatch.init(BugwatchOptions.builder()
 *     .apiKey("your-api-key")
 *     .environment("production")
 *     .build());
 * }</pre>
 * <p>
 * Then capture errors anywhere:
 * <pre>{@code
 * try {
 *     riskyOperation();
 * } catch (Exception e) {
 *     Bugwatch.captureException(e);
 * }
 * }</pre>
 */
public final class Bugwatch {

    private static volatile BugwatchClient instance;
    private static final Object LOCK = new Object();

    private Bugwatch() {
    }

    /**
     * Initializes the SDK with the given options. Must be called before any capture methods.
     * Uses double-checked locking to ensure thread-safe singleton initialization.
     *
     * @param options the SDK configuration options
     */
    public static void init(BugwatchOptions options) {
        if (instance == null) {
            synchronized (LOCK) {
                if (instance == null) {
                    instance = new BugwatchClient(options);
                }
            }
        }
    }

    /**
     * Initializes the SDK with a pre-built client. Useful for testing.
     *
     * @param client the client instance
     */
    public static void init(BugwatchClient client) {
        synchronized (LOCK) {
            instance = client;
        }
    }

    /**
     * Captures an exception and sends it to Bugwatch.
     *
     * @param throwable the throwable to capture
     * @return the event ID, or {@code null} if the SDK is not initialized
     */
    public static String captureException(Throwable throwable) {
        BugwatchClient client = instance;
        if (client == null) {
            System.err.println("[Bugwatch] SDK not initialized. Call Bugwatch.init() first.");
            return null;
        }
        return client.captureException(throwable);
    }

    /**
     * Captures a message and sends it to Bugwatch.
     *
     * @param message the message to capture
     * @param level   the severity level
     * @return the event ID, or {@code null} if the SDK is not initialized
     */
    public static String captureMessage(String message, Level level) {
        BugwatchClient client = instance;
        if (client == null) {
            System.err.println("[Bugwatch] SDK not initialized. Call Bugwatch.init() first.");
            return null;
        }
        return client.captureMessage(message, level);
    }

    /**
     * Flushes any buffered events.
     */
    public static void flush() {
        BugwatchClient client = instance;
        if (client != null) {
            client.flush();
        }
    }

    /**
     * Closes the SDK and releases resources.
     */
    public static void close() {
        synchronized (LOCK) {
            if (instance != null) {
                instance.close();
                instance = null;
            }
        }
    }

    /**
     * Returns the underlying client instance, or {@code null} if not initialized.
     */
    static BugwatchClient getClient() {
        return instance;
    }
}
